provisioning/service/src/Config/EnrollmentGroup.cs

```diff
@@ -0,0 +1,313 @@
+        /// <param name="iotHubHostName">the <code>string</code> with the target IoTHub name. This is optional and can be <code>null</code> or empty.</param>
+        /// <param name="initialTwinState">the <see cref="TwinState"/> with the initial Twin condition. This is optional and can be <code>null</code>.</param>
+        /// <param name="provisioningStatus">the <see cref="ProvisioningStatus"/> that determine the initial status of the device. This is optional and can be <code>null</code>.</param>
+        /// <param name="createdDateTimeUtc">the <code>DateTime</code> with the date and time that the enrollment was created. This is optional and can be <code>null</code>.</param>
+        /// <param name="lastUpdatedDateTimeUtc">the <code>DateTime</code> with the date and time that the enrollment was updated. This is optional and can be <code>null</code>.</param>
+        /// <param name="eTag">the <code>string</code> with the eTag that identify the correct instance of the enrollment in the service. It cannot be <code>null</code> or empty.</param>
+        /// <exception cref="ProvisioningServiceClientException">if the received JSON is invalid.</exception>
+        [JsonConstructor]
+        public EnrollmentGroup(
+            string enrollmentGroupId,
+            AttestationMechanism attestation,
+            string iotHubHostName,
+            TwinState initialTwinState,
+            ProvisioningStatus provisioningStatus,
+            DateTime createdDateTimeUtc,
+            DateTime lastUpdatedDateTimeUtc,
+            string eTag)
+        {
+            /* SRS_ENROLLMENT_GROUP_21_003: [The constructor shall throws ProvisioningServiceClientException if one of the 
+                                                    provided parameters in JSON is not valid.] */
+            if (string.IsNullOrWhiteSpace(eTag))
+            {
+                throw new ProvisioningServiceClientException("Service respond an enrollmentGroup without eTag.");
+            }
+            if(attestation == null)
+            {
+                throw new ProvisioningServiceClientException("Service respond an enrollmentGroup without attestation.");
+            }
+            try
+            {
+                /* SRS_ENROLLMENT_GROUP_21_004: [The constructor shall store all parameters in the JSON.] */
+                EnrollmentGroupId = enrollmentGroupId;
+                Attestation = attestation.GetAttestation();
+                IotHubHostName = iotHubHostName;
+                InitialTwinState = initialTwinState;
+                ProvisioningStatus = provisioningStatus;
+                CreatedDateTimeUtc = createdDateTimeUtc;
+                LastUpdatedDateTimeUtc = lastUpdatedDateTimeUtc;
+                ETag = eTag;
+            }
+            catch (ArgumentException e)
+            {
+                throw new ProvisioningServiceClientException(e);
```

Comment:
I don't think we should convert the exception here. That seems to be the job of the protocol, not the parser.

Reply:
We need to convert because we are using the object setters. When the user set something wrong, it should throw ArgumentException, but when the problem is the JSON received from the service, ProvisioningServiceClientException is the write one.